Provide better debug info

almarklein opened the pull request:
Closes  #23

* [x] Variable names are set consistently.
* [x] Variable names are traced when indexing/getattr.
* [x] Constants get a "name" (i.e. their value).
* [x] Move source filename and linenr along with the bytecodes.
* [x] Use source lines and variable names for better error messages in py compiler.
* [x] Use source lines and variable names for better error messages in SpirV compiler.
* [x] Annotate source lines and/or variable names in SpirV.
* [x] Discover and chase bug #54
* [x] For SpirV only annotate variables and constants, as emitting on co_store_name is glitchy because as it is now, an id can be associated with multiple names.

almarklein commented:
As an example. Consider this shader:
```py
from pyshader import Array, f32, i32, ivec3

def compute_shader(
    index: ("input", "GlobalInvocationId", ivec3),
    data1: ("buffer", 0, Array(f32)),
    data2: ("buffer", 1, Array(i32)),
):
    i = index.x
    arr = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
    for j in range(10):
        arr[j] = arr[j] * 2.0
    val_f = f32(arr[i])
    val_i = i32(arr[i])
    data1[i] = val_f + 1.0
    data2[i] = val_i + 1

```

It has an error; when you try to compile this to SpirV it fails. Before this PR, you'd get:
```
Traceback (most recent call last):
  File "<tmp 2>", line 18, in <module>
    x.to_spirv()
  File "c:\dev\pylib\pyshader\pyshader\_module.py", line 39, in to_spirv
    gen.convert(self._bytecode)
  File "c:\dev\pylib\pyshader\pyshader\_generator_base.py", line 196, in convert
    self._convert(input)
  File "c:\dev\pylib\pyshader\pyshader\_generator_bc.py", line 130, in _convert
    method(*args)
  File "c:\dev\pylib\pyshader\pyshader\_generator_bc.py", line 1067, in co_binary_op
    raise ShaderError(
pyshader._coreutils.ShaderError: Cannot ADD two values with different (sub)types: i32 and f32
```

Not very useful - good luck finding your bug in a 100-line shader.

But with this new PR:
```
Traceback (most recent call last):
  File "<tmp 2>", line 18, in <module>
    x.to_spirv()
  File "c:\dev\pylib\pyshader\pyshader\_module.py", line 39, in to_spirv
    gen.convert(self._bytecode)
  File "c:\dev\pylib\pyshader\pyshader\_generator_base.py", line 196, in convert
    self._convert(input)
  File "c:\dev\pylib\pyshader\pyshader\_generator_bc.py", line 130, in _convert
    method(*args)
  File "c:\dev\pylib\pyshader\pyshader\_generator_bc.py", line 1067, in co_binary_op
    raise ShaderError(
pyshader._coreutils.ShaderError: 
  Source file "<tmp 2>", line 12
  Related variables: arr[..], 2.0
  Cannot MUL two values with different (sub)types: i32 and f32
```

windelbouwman commented on pyshader/_generator_bc.py:
```diff
@@ -135,8 +136,46 @@ def _get_label_id(self, label_value):
+        variables.
+        """
+        filename = getattr(self, "_src_filename", "")
+        linenr = getattr(self, "_src_linenr", 0)  # first line is 1 (not 0)
```
arguably you could use the name `row` here instead of `linenr`, its a shorter name and is easier to pronounce?

almarklein replied:
Sure!

almarklein replied:
No, wait. I think the `filename` plus `linenr` combo is more common and thus easier to recognize than using `row`.

almarklein replied:
Though sometimes `lineno` is used ;)

windelbouwman commented on pyshader/_generator_bc.py:
```diff
@@ -188,7 +227,9 @@ def co_entrypoint(self, name, shader_type, execution_modes):
     def co_func_end(self):
         if self._current_branch is not self._root_branch:
-            raise RuntimeError("Function ends with unresolved sub-branches!")
+            raise RuntimeError(
```
Maybe create a helper function called `error` which takes the message, and optionally some variables and raises the `ShaderError`?

almarklein replied:
I considered that, but I don't like the parrern of calling a function to raise an error, in part because it ads an extra layer in the stack trace.

Korijn commented:
Would it be possible to also include the verbatim source line in the error message?

almarklein commented:
Ok, I seem to have been sloppy in creating that example. The source line *is* normally included, but I ran this interactively, so the filename is not something that can be used to load the source from. Also, in copying the source, I removed some code, and therefore the linenumbers do not match.

Anyway, the example is now updated.

almarklein commented:
(it does show how the code falls back nicely when the filename is not `os.path.isfile` ;) )

Korijn commented:
Awesome

almarklein commented:
oh right ..... the bytecode now includes the source filename, which is different in CI than on my machine, so all tests fail on a bytecode mismatch ...
